entry/entry_io.go

```diff
@@ -1,86 +1,238 @@
+			if !entry.IsValid() {
+				return
+			}
+
+			ts := entry.GetClock().GetTime()
+
+			// Update min/max clocks
+			if maxClock < ts {
+				maxClock = ts
+			}
+
+			if len(result) > 0 {
+				if ts := result[len(result)-1].GetClock().GetTime(); ts < minClock {
+					minClock = ts
+				}
+			} else {
+				minClock = maxClock
+			}
+
+			isLater := len(result) >= length && ts >= minClock
+			// const calculateIndex = (idx) => maxClock - ts + ((idx + 1) * idx)
+
+			// Add the entry to the results if
+			// 1) we're fetching all entries
+			// 2) results is not filled yet
+			// the clock of the entry is later than current known minimum clock time
+			if length < 0 || len(result) < length || isLater {
+				result = append(result, entry)
+				cache[hash] = true
+
+				if options.ProgressChan != nil {
+					options.ProgressChan <- entry
+				}
+
+			}
+
+			if length < 0 {
+				// If we're fetching all entries (length === -1), adds nexts and refs to the queue
+				for i, h := range entry.GetNext() {
+					addToLoadingQueue(h, i)
+				}
+
+				for i, h := range entry.GetRefs() {
+					addToLoadingQueue(h, i)
+				}
+			} else {
+				// If we're fetching entries up to certain length,
+				// fetch the next if result is filled up, to make sure we "check"
+				// the next entry if its clock is later than what we have in the result
+				if _, ok := cache[entry.GetHash()]; len(result) < length || ts > minClock || ts == minClock && !ok {
+					for _, h := range entry.GetNext() {
+						addToLoadingQueue(h, maxClock-ts)
+					}
+				}
+				if len(result)+len(entry.GetRefs()) <= length {
+					for i, h := range entry.GetRefs() {
+						addToLoadingQueue(h, maxClock-ts+((i+1)*i))
+					}
+				}
+			}
+		}
+
+		// Load the entry
+		entry, err := FromMultihash(ctx, ipfs, hash, options.Provider)
 		if err != nil {
-			fmt.Printf("unable to fetch entry %s, %+v\n", hash, err)
+			// TODO: log
 			return
 		}
 
-		entry.Hash = hash
+		// Add it to the results
+		addToResults(entry)
+	}
+
+	// Add entries to exclude from processing to the cache before we start
+	// Add entries that we don't need to fetch to the "cache"
+	for _, e := range options.Exclude {
+		cache[e.GetHash()] = true
+	}
 
-		if entry.IsValid() {
-			addToResults(entry)
+	loadingQueueHasItems := func() bool {
+		for _, s := range loadingQueue {
+			if len(s) > 0 {
+				return true
+			}
 		}
+
+		return false
 	}
 
-	for shouldFetchMore() {
-		fetchEntry()
+	go func() {
+		// Does the loading queue have more to process?
+		for loadingQueueHasItems() {
+			if running < concurrency {
+				nexts := getNextFromQueue(concurrency)
+				running += len(nexts)
+				for _, n := range nexts {
+					fetchEntry(n)
+				}
+
+				running -= len(nexts)
+			}
+		}
+		done <- true
+	}()
+
+	// Resolve the promise after a timeout (if given) in order to
+	// not get stuck loading a block that is unreachable
+	if options.Timeout != 0 {
+		select {
+		case <-time.After(options.Timeout):
+			return result
```

Comment:
it worths having a `cancel()` call here (or in a defer), for now, you should be able to only add a TODO comment as it's not a big issue